include/boost/hana/functional/overload_linearly.hpp

```diff
@@ -96,6 +96,11 @@ BOOST_HANA_NAMESPACE_BEGIN
             return (*this)(static_cast<F&&>(f),
                     (*this)(static_cast<G&&>(g), static_cast<H&&>(h)...));
         }
+
+        template <typename F>
+        constexpr decltype(auto) operator()(F&& f) const {
```

Comment:
What I mean is that instead of `decltype(auto)`, it should be `typename detail::decay<F>::type` according to my argument above (which you could argue against though).